File: app/src/main/java/com/zimuzhedang/subtitleblocker/domain/OverlayState.java
package com.zimuzhedang.subtitleblocker.domain;

/**
 * 悬浮窗状态的不可变模型类。
 * 包含位置、大小、可见性、配置项以及交互状态（拖拽/缩放）。
 *
 * @author Trae
 * @since 2026-01-30
 */
public final class OverlayState {
    /** 悬浮窗宽度 (像素) */
    public final int widthPx;
    /** 悬浮窗高度 (像素) */
    public final int heightPx;
    /** 悬浮窗左上角 X 坐标 (像素) */
    public final int xPx;
    /** 悬浮窗左上角 Y 坐标 (像素) */
    public final int yPx;
    /** 悬浮窗是否可见 */
    public final boolean visible;
    /** 关闭按钮在悬浮窗上的位置 */
    public final CloseButtonPosition closeButtonPosition;
    /** 是否开启交互提示音 */
    public final boolean soundEnabled;
    /** 是否开启常驻通知保持活跃 */
    public final boolean keepAliveEnabled;
    /** 是否正在被拖拽 */
    public final boolean isDragging;
    /** 是否正在被缩放 */
    public final boolean isResizing;

    /**
     * 全参数构造函数。
     */
    public OverlayState(
            int widthPx,
            int heightPx,
            int xPx,
            int yPx,
            boolean visible,
            CloseButtonPosition closeButtonPosition,
            boolean soundEnabled,
            boolean keepAliveEnabled,
            boolean isDragging,
            boolean isResizing
    ) {
        this.widthPx = widthPx;
        this.heightPx = heightPx;
        this.xPx = xPx;
        this.yPx = yPx;
        this.visible = visible;
        this.closeButtonPosition = closeButtonPosition;
        this.soundEnabled = soundEnabled;
        this.keepAliveEnabled = keepAliveEnabled;
        this.isDragging = isDragging;
        this.isResizing = isResizing;
    }

    /**
     * 更新位置并返回新状态。
     *
     * @param xPx 新的 X 坐标
     * @param yPx 新的 Y 坐标
     * @return 新的状态对象
     */
    public OverlayState withPosition(int xPx, int yPx) {
        return new OverlayState(
                widthPx,
                heightPx,
                xPx,
                yPx,
                visible,
                closeButtonPosition,
                soundEnabled,
                keepAliveEnabled,
                isDragging,
                isResizing
        );
    }

    /**
     * 更新尺寸并返回新状态。
     *
     * @param widthPx 新的宽度
     * @param heightPx 新的高度
     * @return 新的状态对象
     */
    public OverlayState withSize(int widthPx, int heightPx) {
        return new OverlayState(
                widthPx,
                heightPx,
                xPx,
                yPx,
                visible,
                closeButtonPosition,
                soundEnabled,
                keepAliveEnabled,
                isDragging,
                isResizing
        );
    }

    /**
     * 更新可见性并返回新状态。
     *
     * @param visible 是否可见
     * @return 新的状态对象
     */
    public OverlayState withVisibility(boolean visible) {
        return new OverlayState(
                widthPx,
                heightPx,
                xPx,
                yPx,
                visible,
                closeButtonPosition,
                soundEnabled,
                keepAliveEnabled,
                isDragging,
                isResizing
        );
    }

    /**
     * 更新关闭按钮位置并返回新状态。
     *
     * @param position 新的按钮位置
     * @return 新的状态对象
     */
    public OverlayState withCloseButtonPosition(CloseButtonPosition position) {
        return new OverlayState(
                widthPx,
                heightPx,
                xPx,
                yPx,
                visible,
                position,
                soundEnabled,
                keepAliveEnabled,
                isDragging,
                isResizing
        );
    }

    /**
     * 更新音效开关并返回新状态。
     *
     * @param enabled 是否启用音效
     * @return 新的状态对象
     */
    public OverlayState withSoundEnabled(boolean enabled) {
        return new OverlayState(
                widthPx,
                heightPx,
                xPx,
                yPx,
                visible,
                closeButtonPosition,
                enabled,
                keepAliveEnabled,
                isDragging,
                isResizing
        );
    }

    /**
     * 更新常驻后台开关并返回新状态。
     *
     * @param enabled 是否启用常驻后台
     * @return 新的状态对象
     */
    public OverlayState withKeepAliveEnabled(boolean enabled) {
        return new OverlayState(
                widthPx,
                heightPx,
                xPx,
                yPx,
                visible,
                closeButtonPosition,
                soundEnabled,
                enabled,
                isDragging,
                isResizing
        );
    }

    /**
     * 更新拖拽状态并返回新状态。
     *
     * @param dragging 是否正在拖拽
     * @return 新的状态对象
     */
    public OverlayState withDragging(boolean dragging) {
        return new OverlayState(
                widthPx,
                heightPx,
                xPx,
                yPx,
                visible,
                closeButtonPosition,
                soundEnabled,
                keepAliveEnabled,
                dragging,
                isResizing
        );
    }

    /**
     * 更新缩放状态并返回新状态。
     *
     * @param resizing 是否正在缩放
     * @return 新的状态对象
     */
    public OverlayState withResizing(boolean resizing) {
        return new OverlayState(
                widthPx,
                heightPx,
                xPx,
                yPx,
                visible,
                closeButtonPosition,
                soundEnabled,
                keepAliveEnabled,
                isDragging,
                resizing
        );
    }
}

File: Flutter_Project/lib/features/overlay/domain/overlay_state.dart
import '../../../core/contracts/business_constants.dart';
import '../../../shared/models/close_button_position.dart';
import '../../settings/domain/settings.dart';
import 'overlay_geometry.dart';

final class OverlayState {
  const OverlayState({
    this.widthPx = OverlayBusinessConstants.defaultOverlayWidthDp,
    this.heightPx = OverlayBusinessConstants.defaultOverlayHeightDp,
    this.xPx = 0,
    this.yPx = 0,
    this.visible = false,
    this.closeButtonPosition = SettingsDefaults.closeButtonPosition,
    this.soundEnabled = SettingsDefaults.soundEnabled,
    this.keepAliveEnabled = SettingsDefaults.keepAliveEnabled,
    this.transparencyToggleEnabled = SettingsDefaults.transparencyToggleEnabled,
    this.transparentMode = false,
    this.isDragging = false,
    this.isResizing = false,
    this.isMinimized = false,
  });

  factory OverlayState.fromSettings(
    Settings settings, {
    int widthPx = OverlayBusinessConstants.defaultOverlayWidthDp,
    int heightPx = OverlayBusinessConstants.defaultOverlayHeightDp,
    int xPx = 0,
    int yPx = 0,
    bool visible = false,
  }) {
    return OverlayState(
      widthPx: widthPx,
      heightPx: heightPx,
      xPx: xPx,
      yPx: yPx,
      visible: visible,
      closeButtonPosition: settings.closeButtonPosition,
      soundEnabled: settings.soundEnabled,
      keepAliveEnabled: settings.keepAliveEnabled,
      transparencyToggleEnabled: settings.transparencyToggleEnabled,
      transparentMode: false,
      isDragging: false,
      isResizing: false,
      isMinimized: false,
    );
  }

  final int widthPx;
  final int heightPx;
  final int xPx;
  final int yPx;
  final bool visible;
  final CloseButtonPosition closeButtonPosition;
  final bool soundEnabled;
  final bool keepAliveEnabled;
  final bool transparencyToggleEnabled;
  final bool transparentMode;
  final bool isDragging;
  final bool isResizing;
  final bool isMinimized;

  OverlaySize get size => OverlaySize(widthPx: widthPx, heightPx: heightPx);

  OverlayPosition get position => OverlayPosition(xPx: xPx, yPx: yPx);

  bool get isVisible => visible;

  bool get isTransparent => transparentMode;

  OverlayState copyWith({
    int? widthPx,
    int? heightPx,
    int? xPx,
    int? yPx,
    bool? visible,
    CloseButtonPosition? closeButtonPosition,
    bool? soundEnabled,
    bool? keepAliveEnabled,
    bool? transparencyToggleEnabled,
    bool? transparentMode,
    bool? isDragging,
    bool? isResizing,
    bool? isMinimized,
  }) {
    return OverlayState(
      widthPx: widthPx ?? this.widthPx,
      heightPx: heightPx ?? this.heightPx,
      xPx: xPx ?? this.xPx,
      yPx: yPx ?? this.yPx,
      visible: visible ?? this.visible,
      closeButtonPosition: closeButtonPosition ?? this.closeButtonPosition,
      soundEnabled: soundEnabled ?? this.soundEnabled,
      keepAliveEnabled: keepAliveEnabled ?? this.keepAliveEnabled,
      transparencyToggleEnabled:
          transparencyToggleEnabled ?? this.transparencyToggleEnabled,
      transparentMode: transparentMode ?? this.transparentMode,
      isDragging: isDragging ?? this.isDragging,
      isResizing: isResizing ?? this.isResizing,
      isMinimized: isMinimized ?? this.isMinimized,
    );
  }

  OverlayState withMinimized(bool minimized) {
    return copyWith(isMinimized: minimized);
  }

  OverlayState withPosition(int xPx, int yPx) {
    return copyWith(xPx: xPx, yPx: yPx);
  }

  OverlayState withSize(int widthPx, int heightPx) {
    return copyWith(widthPx: widthPx, heightPx: heightPx);
  }

  OverlayState withVisibility(bool visible) {
    return copyWith(visible: visible);
  }

  OverlayState withCloseButtonPosition(CloseButtonPosition position) {
    return copyWith(closeButtonPosition: position);
  }

  OverlayState withSoundEnabled(bool enabled) {
    return copyWith(soundEnabled: enabled);
  }

  OverlayState withKeepAliveEnabled(bool enabled) {
    return copyWith(keepAliveEnabled: enabled);
  }

  OverlayState withTransparencyToggleEnabled(bool enabled) {
    return copyWith(transparencyToggleEnabled: enabled);
  }

  OverlayState withTransparentMode(bool enabled) {
    return copyWith(transparentMode: enabled);
  }

  OverlayState withDragging(bool dragging) {
    return copyWith(isDragging: dragging);
  }

  OverlayState withResizing(bool resizing) {
    return copyWith(isResizing: resizing);
  }

  @override
  bool operator ==(Object other) {
    return identical(this, other) ||
        other is OverlayState &&
            other.widthPx == widthPx &&
            other.heightPx == heightPx &&
            other.xPx == xPx &&
            other.yPx == yPx &&
            other.visible == visible &&
            other.closeButtonPosition == closeButtonPosition &&
            other.soundEnabled == soundEnabled &&
            other.keepAliveEnabled == keepAliveEnabled &&
            other.transparencyToggleEnabled == transparencyToggleEnabled &&
            other.transparentMode == transparentMode &&
            other.isDragging == isDragging &&
            other.isResizing == isResizing &&
            other.isMinimized == isMinimized;
  }

  @override
  int get hashCode {
    return Object.hash(
      widthPx,
      heightPx,
      xPx,
      yPx,
      visible,
      closeButtonPosition,
      soundEnabled,
      keepAliveEnabled,
      transparencyToggleEnabled,
      transparentMode,
      isDragging,
      isResizing,
      isMinimized,
    );
  }
}
